### PR TITLE
onchain metadata support for multi delegate

### PR DESCRIPTION
**Pull request by mdtanrikulu**

### CODE REVIEW

File: contracts/ERC20MultiDelegate.sol [hunk 1]
```diff
@@ -108,15 +117,22 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
             transferIndex < Math.max(sourcesLength, targetsLength);
+
         ) {
             address source = address(0);
             address target = address(0);
             if (transferIndex < sourcesLength) {
-                require((sources[transferIndex] >> 160) == 0, "Upper 96 bits of source uint256 must be zero");
+                require(
+                    (sources[transferIndex] >> 160) == 0,
+                    "Upper 96 bits of source uint256 must be zero"
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:127]: These should be custom errors.

File: contracts/ERC20MultiDelegate.sol [hunk 2]
```diff
@@ -108,15 +117,22 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
         for (
             uint transferIndex = 0;
             transferIndex < Math.max(sourcesLength, targetsLength);
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:119]: IIRC one of the gas optimisation suggestions was to calculate the `max` outside the loop.

File: contracts/ERC20MultiDelegate.sol [hunk 3]
```diff
@@ -173,9 +191,63 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
         require(token.transferFrom(proxyAddressFrom, msg.sender, amount));
     }
 
-    function setUri(string memory uri) external onlyOwner {
-        _setURI(uri);
-        emit MetadataURIUpdated(uri);
+    function setUri(string memory /*uri*/) external view onlyOwner {
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:194]: Can we just remove this instead?

File: contracts/ERC20MultiDelegate.sol [hunk 4]
```diff
@@ -173,9 +191,63 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
+        revert NotImplemented();
+    }
+
+    function tokenURI(uint256 tokenId) public view returns (string memory) {
+        string memory addressString = Strings.toHexString(uint160(tokenId), 20);
+        string memory reversedName = string.concat(
+            this.trim(addressString, 2, bytes(addressString).length),
+            ".addr.reverse"
+        );
+        (bytes memory encodedReversedName, ) = reversedName.dnsEncodeName();
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:204]:
```suggestion
        bytes memory encodedReversedName = bytes.concat("\x28", addressString, "\x04addr\x07reverse\x00");
```

File: contracts/ERC20MultiDelegate.sol [hunk 5]
```diff
@@ -173,9 +191,63 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
+        string memory resolvedName;
+        try metadataResolver.reverse(encodedReversedName) returns (
+            string memory _resolvedName,
+            address,
+            address,
+            address
+        ) {
+            resolvedName = _resolvedName;
+        } catch {
+            return "";
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:215]: A failure to get the primary name shouldn't result in all the metadata being empty. Perhaps just return the hex-encoded address instead?

File: contracts/ERC20MultiDelegate.sol [hunk 6]
```diff
@@ -173,9 +191,63 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
+        }
+
+        (bytes memory encodedName, bytes32 namehash) = resolvedName
+            .dnsEncodeName();
+        bytes memory data = abi.encodeWithSignature(
+            "text(bytes32,string)",
+            [namehash, "avatar"]
+        );
+        (bytes memory result, ) = metadataResolver.resolve(encodedName, data);
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:224]: This also needs a try block.

File: contracts/ERC20MultiDelegate.sol [hunk 7]
```diff
@@ -173,9 +191,63 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
+                    '", "description": "This NFT is a proof for your ENS delegation strategy.", "image": "',
+                    imageUri,
+                    '"}'
+                )
+            )
+        );
+        return string.concat("data:application/json;base64,", json);
+    }
+
+    function trim(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:245]: 009ea33df7aca6f9c65c51ae10df081d

Copying the address-to-hex implementation into /utils/HexUtils.sol, and removing the 0x prefix, would remove the need for this function.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: contracts/ERC20MultiDelegate.sol (hunk 2 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: moved it

File: contracts/ERC20MultiDelegate.sol (hunk 7 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: added a new HexUtils method, which converts address to hex string without hex prefix. ptal if that's okay.

File: contracts/ERC20MultiDelegate.sol [hunk 8]
```diff
@@ -104,19 +115,19 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
-        for (
-            uint transferIndex = 0;
-            transferIndex < Math.max(sourcesLength, targetsLength);
-        ) {
+        for (uint transferIndex = 0; transferIndex < maxLength; ) {
             address source = address(0);
             address target = address(0);
             if (transferIndex < sourcesLength) {
-                require((sources[transferIndex] >> 160) == 0, "Upper 96 bits of source uint256 must be zero");
+                if ((sources[transferIndex] >> 160) != 0)
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:124]: Please run lint, and always use braces on if/else blocks.

File: contracts/ERC20MultiDelegate.sol [hunk 9]
```diff
@@ -104,19 +115,19 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
+                    revert Upper96BitsNotZero();
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:125]: This should probably be less implementation-specific, like `InvalidDelegateAddress`.

File: contracts/ERC20MultiDelegate.sol [hunk 10]
```diff
@@ -173,9 +186,72 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
...
+            } catch {}
+        } else {
+            resolvedName = hexAddress;
+        }
+
+        string memory json = Base64.encode(
+            bytes(
+                string.concat(
+                    '{"name": "',
+                    resolvedName,
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:244]: This and `imageUri` need to be escaped.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: contracts/ERC20MultiDelegate.sol (hunk 10 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:246]: Do you think double quotes and backslash escaping will be enough?

**Review by Arachnid — APPROVED** [review]

**Pull request merged**